src/directLineStreaming.ts

```diff
@@ -0,0 +1,345 @@
+    }
+
+    let activity = activitySet.activities[0];
+    activity.attachments = attachments;
+
+    if (this.connectionStatus$.value === ConnectionStatus.Online) {
+      this.subscriber.next(activity);
+    } else {
+      this.activityQueue.push(activity);
+    }
+
+    let r = new BFSE.StreamingResponse();
+    r.statusCode = 200;
+    return r;
+  }
+
+  public flush() {
+    this.connectionStatus$.subscribe(cs => { })
+    this.activityQueue.forEach((a) => this.subscriber.next(a));
+    this.activityQueue = [];
+  }
+}
+
+export class DirectLineStreaming implements IBotConnection {
+  public connectionStatus$ = new BehaviorSubject(ConnectionStatus.Uninitialized);
+  public activity$: Observable<Activity>;
+
+  private activitySubscriber: Subscriber<Activity>;
+  private theStreamHandler: StreamHandler;
+  private parentDirectLine: DirectLine;
+
+  private retryCount = MAX_RETRY_COUNT;
+
+  private domain = "https://directline.botframework.com/v3/directline";
+
+  private conversationId: string;
+  private token: string;
+  private streamConnection: BFSE.WebSocketClient;
+
+  private _botAgent = '';
+
+  constructor(options: DirectLineOptions, parentDirectLine: DirectLine) {
+    this.token = options.secret || options.token;
+
+    if (options.token) {
+      this.refreshToken();
+    }
+
+    if (options.domain) {
+      this.domain = options.domain;
+    }
+
+    if (options.conversationId) {
+      this.conversationId = options.conversationId;
+    }
+
+    this.parentDirectLine = parentDirectLine;
+
+    this._botAgent = this.getBotAgent(options.botAgent);
+
+    this.activity$ = this.streamingWebSocketActivity$().share();
+      this.parentDirectLine.activity$ = this.activity$;
+      this.parentDirectLine.connectionStatus$ = this.connectionStatus$;
+      this.parentDirectLine.postActivity = this.postActivity.bind(this);
+      this.parentDirectLine.end = this.end.bind(this);
+  }
+
+  public reconnect(conversation: Conversation) {
+    this.conversationId = conversation.conversationId;
+    this.token = conversation.token;
+    this.connectAsync().then(_ => _);
+  }
+
+  end() {
+    this.connectionStatus$.next(ConnectionStatus.Ended);
+    this.streamConnection.disconnect();
+  }
+
+  private commonHeaders() {
+    return {
+      "Authorization": `Bearer ${this.token}`,
+      "x-ms-bot-agent": this._botAgent
+    };
+  }
+
+  private getBotAgent(customAgent: string = ''): string {
+    let clientAgent = 'directlinejs'
+
+    if (customAgent) {
+      clientAgent += `; ${customAgent}`
+    }
+
+    return `${DIRECT_LINE_VERSION} (${clientAgent})`;
+  }
+
+  private async refreshToken(firstCall = true, retryCount = 0) {
+    if (firstCall) {
+      setTimeout(async () => await this.refreshToken(false, 0), refreshTokenInterval);
+      return;
+    }
+
+    if (this.connectionStatus$.value === ConnectionStatus.Ended) {
+      return;
+    }
+
+    await this.waitUntilOnline();
+
+    Observable.ajax({
```

Comment:
This doesn't need to use `Observable.ajax`. Can you change it to `fetch` with a while-loop to simplify the logic?

```js
while (!ended && retries > 0) {
  try {
    await fetch(...);
  } catch (err) {
    console.warn('failed to refresh token');
    retries--;
    // do some sleep to prevent spamming the server
    return;
  }

  retries = DEFAULT_RETRIES;
  // sleep for refreshTokenInterval
}
```

Reply:
#262.